### ISSUE TITLE
orchestrator-client and orchestrator command line usage differences

### ISSUE THREAD

**Issue by sjmudd** [openark/orchestrator#901]
While migrating from using the orchestrator command line tool to orchestrator-client to ensure that all access is done via an https interface I notice some differences:

e.g.
```
$ sudo orchestrator -c topology | head -3
host-2.example.com:3306              [0s,ok,5.7.25-log,rw,ROW,>>,P-GTID]
+ host-1.example.com:3306            [0s,ok,5.7.25-log,ro,ROW,>>,GTID,P-GTID]
  + host-3.example.com:3306          [0s,ok,5.7.26-log,ro,nobinlog,GTID,P-GTID]
```
`sudo` required as my `/etc/orchestrator.conf.json` configuration is only readable by root.

If I want to do the same with `orchestrator-client` I first setup something in `/etc/profile.d/orchestrator-client.sh` which will configure an appropriate setting for `ORCHESTRATOR_API`.

However, now running this I use the same input and get:

```
$ orchestrator-client -c topology
instance|alias must be provided
```
This is not required using the `orchestrator` binary.

The current work-around is this:

```
$ orchestrator-client -c topology -i $(hostname) | head -2
host-2.example.com:3306 [1s,ok,5.7.25-log,rw,ROW,>>,P-GTID]
+ host-1.example.com:3306 [1s,ok,5.7.25-log,ro,ROW,>>,GTID,P-GTID]
```

Several `orchestrator` commands allow the use of an implicit `-i xxxxx` value and I'd like to do the same (for compatibility) in `orchestrator-client`.

What's the best way from your perspective to approach this? On a case by case basis where I notice this functionality is missing, or in a more general manner? I can take a look at providing a patch but wasn't sure your preferred approach and first wanted to report the issue.

**Comment by shlomi-noach** [comment]
Hi @sjmudd , yeah, this point is valid. I'm trying do dig in memory lane and see if there were an intention behind this. 

Only thing is, say I run `orchestrator-client` from a non-MySQL host (e.g. a jump host). If I run some `orchestrator-client -c topology` or `orchestrator-client -c relocate -d some.server`:

- Today, I get an error `instance|alias must be provided`
- If we implicitly assumed `-i` to be the local host, I'd get an error in the form "something went wrong", which would be unclear and weird.

So here's one suggestion: we can make this optional. We already include user-based config via

https://github.com/github/orchestrator/blob/26bdccf050ebda680875496ed9c0658cedccd677/resources/bin/orchestrator-client#L37

We can further apply `instance=` in the same way we apply `orchestrator_api=`:

https://github.com/github/orchestrator/blob/26bdccf050ebda680875496ed9c0658cedccd677/resources/bin/orchestrator-client#L39

So, replace
```shell
instance=
```

with
```shell
instance=${ORCHESTRATOR_INSTANCE:-}
```

And the user (you) will populate `/etc/profile.d/orchestrator-client` with:
```shell
ORCHESTRATOR_INSTANCE=$(hostname)
```

I think that's a straightforward solution. WDYT?

**Comment by sjmudd** [comment]
Hi Shlomi,

Thanks for your prompt response.  So yes, I was trying to figure out the cleanest approach. My main concern had been that I'm not sure which `orchestrator` commands use an _implicit_ `instance` and which don't and given there are a lot of commands doing this everywhere may not be the right thing to do.

Your comment about jump host usage is also valid. It's not clear where you expected to run `orchestrator-client`. I mainly run it from MySQL servers for what are clearly ad-hoc usage but also international automation also may run on local servers. Other automation runs from other servers and may poke at orchestrator and it's aware to use all appropriate settings.

So certainly it may be simpler for me for now to adjust `/etc/profile.d/orchestrator-client` and set `ORCHESTRATOR_INSTANCE` here as long as I can override this from the command line if needed.

If all commands behave the same then this feels fine and probably an easy and clean way to resolve this.

**Comment by shlomi-noach** [comment]
Yeah, so your `puppet` can choose to populate or to not populate this value depending on the host. I think it's reasonable to populate it on mysql boxes and expect it to implicitly indicate the local host.

**Comment by sjmudd** [comment]
So just: https://gist.github.com/sjmudd/bc53c5c67f5a4c6190f7612d2e228a97 ?

**Comment by sjmudd** [comment]
Probably needs a line of docs too, to explain usage.
I can make a PR unless given it's a one line change you prefer just to do it yourself.

**Comment by shlomi-noach** [comment]
```
instance="${ORCHESTRATOR_INSTANCE:-}"
```
which technically is the same thing, I know, but is more explicit about this variable not having to exist.

**Comment by shlomi-noach** [comment]
Please make the PR

**Comment by sjmudd** [comment]
Will do, oh and I might have done the usage  of `:-` differently but if that's the style you prefer then that's fine.

**Comment by shlomi-noach** [comment]
closed by #903
